### PR TITLE
add string utils

### PR DESCRIPTION
**Pull request by yj-qin**
#639

### CODE REVIEW

File: string/string.mbt [hunk 1]
```diff
@@ -118,36 +118,248 @@ pub fn iter(self : String) -> Iter[Char] {
...
+      } else {
+        buf.write_char(c)
+      }
+    },
+  )
+  buf.to_string()
+}
+
+/// Joins an array of string by a seperator.
+pub fn join(strs : Array[String], seperator : String) -> String {
```
Comment (qazxcdswe123) [inline string/string.mbt:354]: FYI: `join` had beed added in https://github.com/moonbitlang/core/pull/597/

**Review by qazxcdswe123 — CHANGES_REQUESTED** [review]

**Comment by qazxcdswe123** [comment]
I think it's better to use KMP search for `index_of`

I've implemented it myself and could you help verify it?

```
fn compute_prefix(pattern : String) -> Array[Int] {
  let m = pattern.length()
  let prefix = Array::make(m, 0)
  let mut k = 0
  for q = 1; q < m; q = q + 1 {
    while k > 0 && pattern[k] != pattern[q] {
      k = prefix[k - 1]
    }
    if pattern[k] == pattern[q] {
      k = k + 1
    }
    prefix[q] = k
  }
  return prefix
}

fn kmp_search(text : String, pattern : String) -> Int {
  let n = text.length()
  let m = pattern.length()
  if m == 0 {
    return 0
  } else if m > n {
    return -1
  }
  let prefix = compute_prefix(pattern)
  let mut q = 0
  for i = 0; i < n; i = i + 1 {
    while q > 0 && pattern[q] != text[i] {
      q = prefix[q - 1]
    }
    if pattern[q] == text[i] {
      q = q + 1
    }
    if q == m {
      return i - m + 1
    }
  }
  return -1
}
```

Thanks!

**Review by qazxcdswe123 — DISMISSED** [review]

**Comment by yj-qin** [comment]
I have implemented a basic brute force matching algorithm, which should be fast when the length of strings and substrings does not exceed a threshold. When the string is longer, it can be switched to KMP or Rabin-Karp algorithm. 
You can do an optimize in another PR.

FYR, Go's string index search:
https://github.com/golang/go/blob/master/src/internal/stringslite/strings.go#L24

@qazxcdswe123

File: string/string.mbt [hunk 2]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
   )
 }
 
+/// Removes the around whitespaces of this string.
+pub fn trim(self : String) -> String {
+  let len = self.length()
+  let mut start = 0
+  let mut end = len
+  while start < end && self[start] <= ' ' {
+    start += 1
```
Comment (bobzhang) [inline string/string.mbt:127]: can you add commment why `self[start] <= '  ' `works, does other language follow the same semantics

File: string/string.mbt [hunk 3]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
...
+  }
+  while start < end && self[end - 1] <= ' ' {
+    end -= 1
```
Comment (bobzhang) [inline string/string.mbt:130]: same here

File: string/string.mbt [hunk 4]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
...
+    if found {
+      return start + 1
+    }
+    i -= 1
+  }
+}
+
+/// Returns true if this string contains a sub string.
+pub fn contains(self : String, str : String) -> Bool {
+  self.index_of(str) > -1
```
Comment (bobzhang) [inline string/string.mbt:236]: ` >= 0`?

File: string/string.mbt [hunk 5]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
...
+    }
+    subs.push(self.substring(~start, ~end))
+    start = end + sep_len
+  }
+  subs
+}
+
+/// Replace the first old string in this string to new string.
+pub fn replace(self : String, old : String, new : String) -> String {
+  let old_idx = self.index_of(old)
```
Comment (bobzhang) [inline string/string.mbt:282]: use the label here? ` (self : String,  ~old: String, ~new : String)`

File: string/string.mbt [hunk 6]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
...
-        return i
+    self.substring(end=old_idx) + new + self.substring(
+      start=old_idx + old.length(),
+    )
+  }
+}
+
+/// Replace all old string in this string to new string.
+pub fn replace_all(self : String, old : String, new : String) -> String {
+  let buf = Buffer::new()
```
Comment (bobzhang) [inline string/string.mbt:294]: same here

File: string/string.mbt [hunk 7]
```diff
@@ -118,36 +118,234 @@ pub fn iter(self : String) -> Iter[Char] {
...
+    }
+    -1
+  }
+}
+
+/// Returns the last index of the sub string.
+pub fn last_index_of(
+  self : String,
+  str : String,
+  ~from : Int = self.length()
```
Comment (bobzhang) [inline string/string.mbt:189]: should `last_index_of` and `index_of` be fused to a single API, e.g,
```
index_of(self : String, str : String, ~rev : Bool = true, ~from = if rev { self.length()} else {0})
```

cc @Guest0x0

**Comment by bobzhang** [comment]
@qazxcdswe123 kmp could be provided via a new API

File: string/string.mbt (hunk 2 above)
Comment (yj-qin) [reply string/string.mbt:127]: The Java JDK was implemented like this. I rethought it, maybe provide `trim`, `trim_left` and `trim_right` with a `trim_chars` parameter, and build `trim_space` on top of them, like Go's api.

**Pull request merged**